### PR TITLE
Run in alpine with glibc/gmp and a proper locale

### PR DESCRIPTION
**Pull request by pbrisbin**
The hGetContents exception which we tried to fix by setting LANG was not
actually fixed. I believe this is because busybux ships with no locale support
and so setting LANG had no effect.

To fix this, I first moved to alpine which is quickly replacing busybox as the
go-to runtime docker base image. The fix then required installation of an
alternate glibc package to define the locale and allowed for installation of
the gmp package instead of copying the library manually.

The image size increases from ~26MB to ~56MB with this change.

This does indicate what looks like an error when checking the binary with ldd:

```
% docker run --user 0 -it --entrypoint /bin/sh --rm codeclimate/codeclimate-shellcheck -c "ldd /home/app/bin/engine"
	/lib64/ld-linux-x86-64.so.2 (0x5621a21c1000)
	librt.so.1 => /lib64/ld-linux-x86-64.so.2 (0x5621a21c1000)
	libutil.so.1 => /lib64/ld-linux-x86-64.so.2 (0x5621a21c1000)
	libdl.so.2 => /lib64/ld-linux-x86-64.so.2 (0x5621a21c1000)
	libgmp.so.10 => /usr/lib/libgmp.so.10 (0x7f1a2838c000)
	libm.so.6 => /lib64/ld-linux-x86-64.so.2 (0x5621a21c1000)
	libpthread.so.0 => /lib64/ld-linux-x86-64.so.2 (0x5621a21c1000)
	libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x7f1a28179000)
	libc.so.6 => /lib64/ld-linux-x86-64.so.2 (0x5621a21c1000)
Error relocating /home/app/bin/engine: pthread_setname_np: symbol not found
```

But the engine appears to run fine (and doesn't exhibit the hGetContents error)

/cc @wfleming

### CODE REVIEW

File: docker/Release.plan [hunk 1]
```diff
@@ -1,9 +1,15 @@
-FROM progrium/busybox
+FROM alpine:3.3
 MAINTAINER Philip Cunningham <hello@filib.io>
 
-ENV LANG en_US.UTF-8
+ADD docker/files/ .
```
Comment (pbrisbin) [inline docker/Release.plan]: Whoops; forgot to commit the `.apk` files being added here. Will do in a moment.

**Comment by pbrisbin** [comment]
Also @filib if you know of a way to actually solve this on busybox, I'd be happy to try a less invasive change. I have a reproduction case we can use to be sure this time.

**Comment by filib** [comment]
> Also @filib if you know of a way to actually solve this on busybox, I'd be happy to try a less invasive change. I have a reproduction case we can use to be sure this time. 

I don't unfortunately. Switching to alpine seems like the right thing to do. 

> Error relocating /home/app/bin/engine: pthread_setname_np: symbol not found

@pbrisbin any idea what this means in practice?

**Comment by pbrisbin** [comment]
>  any idea what this means in practice?

All I can say with certainty is that the engine does still run successfully on the projects I tested it against. I'd imagine this means the code paths followed don't ever try to invoke `pthread_setname_np`, and if it did you'd see a segfault/linker error.

It's worth noting that this build is actually much better than the currently-released engine image:

```
% docker run --user 0 -it --entrypoint /bin/sh --rm us.gcr.io/code_climate/codeclimate-shellcheck:b248 -c "ldd /home/app/bin/engine"                                                            toolbox:master
	/lib64/ld-linux-x86-64.so.2 (0x55e50d395000)                                       
	librt.so.1 => /lib64/ld-linux-x86-64.so.2 (0x55e50d395000)
	libutil.so.1 => /lib64/ld-linux-x86-64.so.2 (0x55e50d395000)
	libdl.so.2 => /lib64/ld-linux-x86-64.so.2 (0x55e50d395000)
	libgmp.so.10 => /usr/lib/libgmp.so.10 (0x7fa15593f000)
	libm.so.6 => /lib64/ld-linux-x86-64.so.2 (0x55e50d395000)
	libpthread.so.0 => /lib64/ld-linux-x86-64.so.2 (0x55e50d395000)
	libgcc_s.so.1 => /lib/libgcc_s.so.1 (0x7fa15572e000)
	libc.so.6 => /lib64/ld-linux-x86-64.so.2 (0x55e50d395000)
Error relocating /usr/lib/libgmp.so.10: __snprintf_chk: symbol not found
Error relocating /usr/lib/libgmp.so.10: __vsnprintf_chk: symbol not found
Error relocating /usr/lib/libgmp.so.10: _obstack_newchunk: symbol not found
Error relocating /usr/lib/libgmp.so.10: __printf_chk: symbol not found
Error relocating /usr/lib/libgmp.so.10: __memset_chk: symbol not found
Error relocating /usr/lib/libgmp.so.10: __fprintf_chk: symbol not found
Error relocating /usr/lib/libgmp.so.10: __vsprintf_chk: symbol not found
Error relocating /usr/lib/libgmp.so.10: obstack_vprintf: symbol not found
Error relocating /home/app/bin/engine: pthread_setname_np: symbol not found
```

I think just no one has thought to check explicitly with `ldd` to this point.

**Comment by wfleming** [comment]
I believe the `ldd` error is down to the fact that `glibc` versions are now different between the build container & the final image container. That's the downside of trying to statically compile with `glibc`: you're never *really* fully statically compiled, so different glibc versions between compilation environment & runtime environment can cause problems. I actually don't recall if I ever checked the `ldd` output on a previous image, so I'm not sure if this is new or not.

If you're both quite confident nothing's going to be trying to spin up threads that just hasn't been triggered somehow yet, then this is probably fine as-is. If you wanted to look into fixing it, the most direct routes would probably be to try and use a build image based off the same alpine version: there's an [alpine-ghc](https://github.com/mitchty/alpine-ghc) image that might be useful.

I'm also wondering if `hlint` really does have to link against `pthread` in the first place if it doesn't appear to be getting used. I'm assuming that was added for because one of the dependencies does use it, but I don't know if it can be built without it?

**Comment by pbrisbin** [comment]
>  I actually don't recall if I ever checked the ldd output on a previous image, so I'm not sure if this is new or not.

I haven't gone back through all the builds, but my test of `b248` shows this and other link errors present in the currently-released version, and it seems to be working well in the wild.

> I'm also wondering if hlint really does have to link against pthread in the first place if it doesn't appear to be getting used

Did you mean shellcheck? I know this engine does some threading stuff that hlint does not.

> If you're both quite confident nothing's going to be trying to spin up threads that just hasn't been triggered somehow yet, then this is probably fine as-is.

That's my general feeling, especially since `b248` is already broken in this way. @filib WDYT?

**Comment by filib** [comment]
> I believe the ldd error is down to the fact that glibc versions are now different between the build container & the final image container. 

@whelming Right! I like your suggestion of using a build container with the same glibc. Makes compete sense. 

> Did you mean shellcheck? I know this engine does some threading stuff that hlint does not.

@pbrisbin this was my worry, I'm using channels for processing results but if this is issue was already present, then I think it's best this gets merged and the problem gets tracked in another issue. 

Merging :tada:

**Comment by filib** [comment]
https://github.com/filib/codeclimate-shellcheck/issues/49

**Pull request merged**